app/scripts/metamask-controller.js

```diff
@@ -375,26 +385,64 @@ module.exports = class MetamaskController extends EventEmitter {
   }
 
   setupProviderConnection (outStream, originDomain) {
-    streamIntoProvider(outStream, this.provider, onRequest, onResponse)
+    // setup json rpc engine stack
+    const engine = new RpcEngine()
+    engine.push(originMiddleware)
+    engine.push(loggerMiddleware)
+    engine.push(createFilterMiddleware({
+      provider: this.provider,
+      blockTracker: this.blockTracker,
+    }))
+    engine.push(createProviderMiddleware({ provider: this.provider }))
+
+    // setup connection
+    const providerStream = createEngineStream({ engine })
+    pump(
+      outStream,
+      providerStream,
+      outStream,
+      (err) => {
+        if (err) console.error(err)
+      }
+    )
+
     // append dapp origin domain to request
-    function onRequest (request) {
-      request.origin = originDomain
+    function originMiddleware (req, res, next, end) {
+      req.origin = originDomain
+      next()
     }
+
     // log rpc activity
-    function onResponse (err, request, response) {
-      if (err) return console.error(err)
-      if (response.error) {
-        console.error('Error in RPC response:\n', response)
+    function loggerMiddleware (req, res, next, end) {
+      next((cb) => {
+        if (res.error) {
+          console.error('Error in RPC response:\n', res)
```

Comment:
Prefer `log.error()`, although I wouldn't block over this, as it was already a `console.error`.